sample: add memfault (sample, memfault SDK, BLE MDS) (#821)
nrfconnect/sdk-nrf-bm: doc/nrf-bm/libraries/bluetooth/services/ble_mds.rst

.. _lib_ble_service_mds:

Memfault Diagnostic Service (MDS)
#################################

.. contents::
   :local:
   :depth: 2

This module implements the `Memfault Diagnostic GATT Service`_ (MDS) over Bluetooth LE on |BMshort|.
It lets an MDS gateway read the Memfault upload URI and authorization from the device, subscribe to data export notifications, and stream Memfault diagnostic chunks to the Memfault cloud.

Overview
********

During initialization, the module registers a vendor-specific GATT service.
The service and characteristic UUIDs follow the `Memfault Diagnostic GATT Service`_ specification and are defined in :file:`include/bm/bluetooth/services/ble_mds.h`.

> Service UUID

Reviewer: Would it make sense to move this after the `Characteristics` section, where the characteristics are described?

SoftDevice BLE events are handled through the observer registered by :c:macro:`BLE_MDS_DEF`.
Heavy Memfault work (packetizer reads and notifications) is intentionally deferred to :c:func:`ble_mds_process` so it runs in main-loop context.
See :ref:`memfault_bm` for Memfault usage rules on bare metal.

Gateway interaction
=====================

A typical MDS gateway session follows these steps:

1. Connect to the peripheral.
2. Read the Device Identifier, Data URI, and Authorization characteristics.
3. Enable notifications on the Data Export Client Characteristic Configuration descriptor (CCCD).
4. Write ``0x01`` to the Data Export characteristic to enable streaming (``0x00`` disables it).
5. Receive chunk notifications until streaming is disabled or the connection drops.
6. Forward received chunks to Memfault using the URI and authorization values.

Only one active MDS subscriber is supported at a time.
If a second central enables notifications while another subscriber is active, the request is ignored.

Configuration
*************

Set the :kconfig:option:`CONFIG_BLE_MDS` Kconfig option to enable the service.

> The service also requires :kconfig:option:`CONFIG_MEMFAULT` and a configured Memfault project key (for example :kconfig:option:`CONFIG_MEMFAULT_NCS_PROJECT_KEY`).

Reviewer 2: ```suggestion
Service also requires the :kconfig:option:`CONFIG_MEMFAULT` Kconfig option and a configured Memfault project key (for example the :kconfig:option:`CONFIG_MEMFAULT_NCS_PROJECT_KEY` Kconfig option).
```

> Service also requires the :kconfig:option:`CONFIG_MEMFAULT` Kconfig option and a configured Memfault project key (for example the :kconfig:option:`CONFIG_MEMFAULT_NCS_PROJECT_KEY` Kconfig option).

Reviewer: ```suggestion
The service also requires enabling the :kconfig:option:`CONFIG_MEMFAULT` Kconfig option and a configured Memfault project key set using the :kconfig:option:`CONFIG_MEMFAULT_NCS_PROJECT_KEY` Kconfig option.
```

Additional options:

* :kconfig:option:`CONFIG_BLE_MDS_DATA_URI_MAX_LEN` – Maximum length of the Data URI characteristic value.
* :kconfig:option:`CONFIG_BLE_MDS_EMPTY_POLL_INTERVAL_MS` – Minimum interval between packetizer polls when no data is available.

> * :kconfig:option:`CONFIG_BLE_MDS_TRIGGER_LOG_COLLECTION` – Periodically snapshot the Memfault RAM log buffer into the packetizer while streaming is active.

Reviewer: I suggest renaming this to better align with the interval Kconfig option below. 
```suggestion
* :kconfig:option:`CONFIG_BLE_MDS_LOG_COLLECTION` – Periodically snapshot the Memfault RAM log buffer into the packetizer while streaming is active.
```

* :kconfig:option:`CONFIG_BLE_MDS_LOG_COLLECTION_INTERVAL_MS` – Interval for log collection when enabled.

Initialization
==============

The service instance is declared using the :c:macro:`BLE_MDS_DEF` macro, specifying the name of the instance.
The macro also registers a SoftDevice BLE observer for the instance.

The service is initialized by calling the :c:func:`ble_mds_init` function.
Use :c:macro:`BLE_MDS_CONFIG_SEC_MODE_DEFAULT` or populate :c:struct:`ble_mds_config` to set GATT security requirements for each characteristic.

After initialization, include the MDS UUID type returned by :c:func:`ble_mds_service_uuid_type` in the advertising UUID list if the service should be discoverable before connection.

Usage
*****

Call :c:func:`ble_mds_process` regularly from the application main loop while a connection may be active.
The function is non-blocking: it returns immediately if no subscriber is streaming, if a notification is already in flight, or if the SoftDevice TX queue is full.

When streaming is enabled and Memfault data is available, the library:

> 1. Optionally triggers Memfault log collection (if :kconfig:option:`CONFIG_BLE_MDS_TRIGGER_LOG_COLLECTION` is enabled).

Reviewer 2: ```suggestion
1. Optionally triggers Memfault log collection (if the :kconfig:option:`CONFIG_BLE_MDS_TRIGGER_LOG_COLLECTION` Kconfig option is enabled).
```

SoftDevice BLE events are handled through the observer registered by :c:macro:`BLE_MDS_DEF`.
Heavy Memfault work (packetizer reads and notifications) is intentionally deferred to :c:func:`ble_mds_process` so it runs in main-loop context.
See :ref:`memfault_bm` for Memfault usage rules on bare metal.

Gateway interaction
=====================

A typical MDS gateway session follows these steps:

1. Connect to the peripheral.
2. Read the Device Identifier, Data URI, and Authorization characteristics.
3. Enable notifications on the Data Export Client Characteristic Configuration descriptor (CCCD).
4. Write ``0x01`` to the Data Export characteristic to enable streaming (``0x00`` disables it).
5. Receive chunk notifications until streaming is disabled or the connection drops.
6. Forward received chunks to Memfault using the URI and authorization values.

Only one active MDS subscriber is supported at a time.
If a second central enables notifications while another subscriber is active, the request is ignored.

Configuration
*************

Set the :kconfig:option:`CONFIG_BLE_MDS` Kconfig option to enable the service.

> Set via :kconfig:option:`CONFIG_MEMFAULT_NCS_PROJECT_KEY`; see :ref:`ble_mds_sample_getting_started_memfault` for setup.

Reviewer: ```suggestion
Set via the :kconfig:option:`CONFIG_MEMFAULT_NCS_PROJECT_KEY` Kconfig option; see :ref:`ble_mds_sample_getting_started_memfault` for setup.
```

* :kconfig:option:`CONFIG_BLE_MDS_LOG_COLLECTION` – Periodically snapshot the Memfault RAM log buffer into the packetizer while streaming is active.
* :kconfig:option:`CONFIG_BLE_MDS_LOG_COLLECTION_INTERVAL_MS` – Interval for log collection when enabled.

Initialization
==============

The service instance is declared using the :c:macro:`BLE_MDS_DEF` macro, specifying the name of the instance.
The macro also registers a SoftDevice BLE observer for the instance.

The service is initialized by calling the :c:func:`ble_mds_init` function.
Use :c:macro:`BLE_MDS_CONFIG_SEC_MODE_DEFAULT` or populate :c:struct:`ble_mds_config` to set GATT security requirements for each characteristic.

After initialization, include the MDS UUID type returned by :c:func:`ble_mds_service_uuid_type` in the advertising UUID list if the service should be discoverable before connection.

Usage
*****

Call :c:func:`ble_mds_process` regularly from the application main loop while a connection may be active.
The function is non-blocking: it returns immediately if no subscriber is streaming, if a notification is already in flight, or if the SoftDevice TX queue is full.

> Recommended main-loop integration” snippet MDS-only:
> 
> .. code-block:: c
> 
>    ble_mds_process(&mds);   /* when CONFIG_BLE_MDS; main loop only */

Reviewer: I don't think this give much added value, it is only one function call. 
```suggestion
```

> Collect and serialize Memfault data per :ref:`memfault_bm` before export.

Reviewer: This result in the following sentence which is a bit strange: 
`Collect and serialize Memfault data per Memfault on Bare Metal before export.`
